### DISCUSSION TITLE
Odd WeakReferenceMessenger error

### DISCUSSION THREAD

**Discussion by CADTraveller** [CommunityToolkit/dotnet#342] [General]
I have been using WeakReferenceMessenger service to communicate between ViewModels. I have nested structure, there is a 'Parent' ViewModel which contains 'Child' ViewModels and I want to get notice of state changes in 'Child' VMs.

In the Parent constructor i register as a receiver:

```
WeakReferenceMessenger.Default.Register<StatusUpdateEventArgs>(this, (r, m) =>
            {
                StatusMessage = m.Message;
            });
```

In the nested Child VM I will publish a message:

```
private void publishStatus(string message)
        {
            WeakReferenceMessenger.Default.Send(new StatusUpdateEventArgs(message));
        }
```

This works as expected in a test Harness project on my machine.

However, when a colleague runs this we are getting this error when sending the message:

_Method not found: 'Boolean System.Runtime.CompilerServices.Unsafe.IsNullRef(!!0 ByRef)_

Anyone recognize this? Since I cannot duplicate on my machine I'm struggling to investigate.

**Comment by Sergio0694** [comment]
That doesn't look right. What NuGet version are you using? What TFM are you using? Are you using .NET 7 preview by any chance? 🤔

**Reply by CADTraveller** [reply]
TFM is 4.7.2, Nuget is 6.3.0

My coworker is running 6.2.1

Definitely NO .NET 7 preview!!  I'm already kind of pushing boundaries here...

**Reply by MichalPetryka** [reply]
Are you using desktop .Net Framework or Mono?

**Reply by CADTraveller** [reply]
Desktop Framework

**Reply by MichalPetryka** [reply]
Are you sure your colleague doesn't use Mono? The only time I've seen a similar issue was there.

**Reply by CADTraveller** [reply]
We do plugins/extensions for AutoCAD and Revit. I double my colleague even knows what Mono is.

Others in the company get the same error, it seems to only work correctly on my machine.

**Reply by CADTraveller** [reply]
I confirmed that we are only running MS .Net runtime.
Could it be related to SDK versions?

**Reply by Sergio0694** [reply]
I suspect it might, as I don't really see how this could be an issue in the MVVM Toolkit.
That said, I'm not really sure what could be causing it, and I can't repro this on my end either 🤔
Can you try running `dotnet --list-sdks` on your machine and your colleague's to see if there's any differences?
Also, are you both using the same VS version? Are you on the same OS version?

**Reply by CADTraveller** [reply]
I have:
3.1.421 [C:\Program Files\dotnet\sdk]
6.0.202 [C:\Program Files\dotnet\sdk]
6.0.203 [C:\Program Files\dotnet\sdk]
6.0.302 [C:\Program Files\dotnet\sdk]
6.0.400-preview.22330.6 [C:\Program Files\dotnet\sdk]
I'm running VS 17.3 Preview 3

Problem machine lists:
3.1.421 [C:\Program Files\dotnet\sdk]
6.0.301 [C:\Program Files\dotnet\sdk]
6.0.302 [C:\Program Files\dotnet\sdk]
Running VS 17.2.5

Both on latest public release of Windows 11

It's tough to see how the VS version would matter, as I had other colleagues get this same error after pulling down the .dll from our TeamCity buildserver.

**Reply by CADTraveller** [reply]
@Sergio0694 I might have finally discovered the cause of this issue.  This came up again recently, my project works fine in VS and on my machine but fails with this same error in production.  A colleague pointed out that it's likely the dependencies are different in the 2 environments, so I spent some time swapping .dlls until I found one which would trigger this error.

![image](https://github.com/CommunityToolkit/dotnet/assets/5366736/18a5da06-a786-4fdd-90e4-a313ebd1fad2)

The version on the left is on my machine, and it sends messages just fine. The version on the right is from our published release, where the error occurs. 

I'm not 100% on this just yet, I still need to update our release and make sure this resolves the issue. That has to be done by our TC admin so there's a bit of a delay.

But it sure looks like this is the culprit.

**Comment by ValeriyLappo** [comment]
I'm also encountered with this problem. using nuget package 8.3.2, but still when using `WeakReferenceMessenger.Send()` method it throws exception with message 

> Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

For more information. If I install RevitPython shell with System.Runtime.CompilerServices.Unsafe.dll version of 4.6.28619.1 and it loads in Revit before my plugin, everything works fine. But if i delete it — got this exception

**Reply by Sergio0694** [reply]
You got some kind of assembly reference mismatch. Try adding an explicit package reference to `System.Runtime.CompilerServices.Unsafe` in your project with the right version. This sometimes happens on older frameworks in some corner case scenarios involving weird transitive dependencies using either packages.config or legacy-style .csproj projects.

**Reply by ValeriyLappo** [reply]
The problem is that i can't do it because i have a lot of other packages with that assembly, including MvvmToolkit. So my VS can't do it. Here an example from output 

> NU1605: Warning As Error: Detected package downgrade: System.Runtime.CompilerServices.Unsafe from 6.0.0 to 4.5.3. Reference the package directly from the project to select a different version. 
> MyProject -> CommunityToolkit.Mvvm 8.3.2 -> System.Runtime.CompilerServices.Unsafe (>= 6.0.0) 
> MyProject -> System.Runtime.CompilerServices.Unsafe (>= 4.5.3)
